source/develop/release-management/features/virt/dedicated-cpu.md

```diff
@@ -0,0 +1,170 @@
+    the requested CPUs are already occupied;
+
+  * it may be impossible to start a VM with dedicated CPUs even though there
+    seems to be enough space in terms of number of used vs. free CPUs.
+
+To workaround this limitation one can start VMs with CPU pinning before VMs
+with dedicated CPU policy. And more generally, the VMs with dedicated CPUs can
+be manually migrated to another host to free the needed CPUs and then possibly
+migrated back.
+
+*TODO*: Tips how to identify candidates for migration?  virsh? Something in UI?
+
+
+## User Experience
+
+## Changes to Engine and Vdsm
+
+The high level overview of the workflow in Engine and Vdsm is following:
+
+1.  When host switches to *UP* state engine queries it to find out which pCPUs
+    are dedicated through a policy or pinned. This is necessary for engine to
+    get a good picture about CPU usage so that it can schedule new VMs
+    properly.
+
+2.  On VM start/migration/etc. Engine consults the policy for dedicating CPUs
+    and schedules the VM on host with enough free CPUs. Engine determines the
+    mapping between physical and virtual CPUs.
+
+3.  Engine produces domain XML that contains properly populated `<vcpupin>`
+    element (similarly to CPU pinning) and stores the policy name in VM
+    metadata.
+
+4.  Vdsm starts the VM and analyzes which CPUs were dedicated. Vdsm updates the
+    pool of shared CPUs and re-configures VMs without any CPU pinning or
+    dedicated CPUs to run on the new set of CPUs from shared pool.
+
+5.  When the VM with CPU pinning or dedicated CPUs is being destroyed, Vdsm
+    updates the pool of shared CPUs. Vdsm then re-configures VMs without any
+    CPU pinning or dedicated CPUs to run on the new set of CPUs from shared
+    pool.
+
+Several changes to the internal API are required. To accomplish step 1. new API
+verb `Host.cpuInfo` needs to be defined that will provide:
+
+  * list of pCPUs dedicated to VMs through CPU pinning or policy
+  * CPU indices and topology information how are sockets, cores and threads
+    organized
+
+Further, Engine will add a new entry into VM metadata called `cpuPolicy` that
+will contain the requested CPU policy. And finally, migration parameters in
+`VM.migrate` will contain a field `cpus` that will contain the CPU mapping for
+the VM on the destination host. This is necessary so that the source VDSM can
+provide a correct domain XML to the destination host.
+
+*TODO*: In what format? Can we use `<vcpupin>` element content or should we use
+something more concise? Alternatively, we can also let destination VDSM parse
+`<vcpupin>` from XML in `VM.migrationCreate` and use libvirt API to pin the CPUs.
+
+## Testing
+
+*TODO*
+
+There are many parts that need to be tested 
+
+feature testing:
+- create and especially migration
+- isolate-threads is respected
+- require-threads excludes unused threads
+
+backward compatibility:
+- cpu pinning
+- migration
+
+
+## Future Work
```

Comment:
auto pinning "existing/pin" feature, to not clash on the same pCPUs

Reply:
the more I think of it, the harder it is for me to find the similarity between what was asked for PIN and this feature (dedicated CPUs).

when you have 1 host with 2 threads and you run 2 VMs with 1 vCPU and policy=PIN then it's similar because you want each VM to be pinned to a different CPU (let's put aside things like VDSM and other system process for simplification)
but let's say that you run a third VM with 1 vCPU and you didn't set it with dedicated policy - its vCPU can be pinned to either of the host threads, right?

so it's similar in a way because in both cases we want the engine to automatically pin vCPUs but if the PIN policy doesn't mean that the pinned vCPU should be pinned exclusively, then it's different and then yeah, vCPUs that were pinned automatically should not be break dedicated CPUs as manual (explicit) pinning

Reply:
Yes, but since the new verb in VDSM - we know which CPUs are taken, so we can create a vCPU pinning according to that. You are right that it's not dedicated, only pinned.
This feature will allow us to PIN to the unused pCPUs. Even without dedicated CPUs. I think it might be also a combination, dedicated+PIN and then it will be also exclusive.

Reply:
> so it's similar in a way because in both cases we want the engine to automatically pin vCPUs but if the PIN policy doesn't mean that the pinned vCPU should be pinned exclusively,

Well, I thought this is rather an unwanted caveat of PIN policy rather than expected behavior.

> then it's different and then yeah, vCPUs that were pinned automatically should not be break dedicated CPUs as manual (explicit) pinning

I don't think I understand this. You cannot expect pinned CPUs (manual or auto) to share with dedicated CPUs. This breaks the expectation of dedicated CPU. So if there is already VM with pinning started another VM with dedicated CPU policy will have to use different CPUs. On the other hand if there is already a VM with dedicated CPU policy already occupying some CPUs than you cannot start a VM with pinning on the same CPUs (which means you cannot start that VM at all).

> Thinking on it a bit more - what will happen if we combine CPU pinning(static) with dedicated?

On a single VM? You cannot, that does not make much sense.

If all of the above is confusing then we can decouple the expectation on sharing (shared/exclusive) from the CPU assignment (static CPU pinning, NUMA auto pinning or automatic assignment what we call *dedicated CPU policy* here). That way you can create all the various combinations of exclusivity and assignment. This would give the user most flexibility but may be little bit difficult to grasp by users. It still does not solve the scheduling problem though -- you cannot run exclusive VMs and shared VMs on same CPUs.

Reply:
> On a single VM? You cannot, that does not make much sense.
Why not? :)
If I want the VM to use specific pCPUs, and also to have them dedicated for my VM. As long they are freed on the host I can't see reason why it won't be possible.

Reply:
> > On a single VM? You cannot, that does not make much sense.
> 
> Why not? :)
> If I want the VM to use specific pCPUs, and also to have them dedicated for my VM. As long they are freed on the host I can't see reason why it won't be possible.

correct me if I'm wrong but I think you both are talking about different things:
@liranr23  meant to have certain vCPUs both pinned and defined as dedicated while @nyoxi  referred to having some of the VMs pinned without being dedicated and some pinned as dedicated
do I get this right?

>If all of the above is confusing then we can decouple the expectation on sharing (shared/exclusive) from the CPU assignment (static CPU pinning, NUMA auto pinning or automatic assignment what we call dedicated CPU policy here). That way you can create all the various combinations of exclusivity and assignment. This would give the user most flexibility but may be little bit difficult to grasp by users. It still does not solve the scheduling problem though -- you cannot run exclusive VMs and shared VMs on same CPUs.

this makes sense to me.
so users will be able to:
1. set pinning explicitly and then set whether they should be shared or exclusive
2. use auto pinning policy - either the 'resize and pin' we have for SAP HANA or 'pin' that doesn't resize or 'dedicated' is similar to 'pin' in the sense that it respects the specified CPU topology but pin it as described above or the other two policies

Reply:
That seems right.
If we give `fixed` CPUs and we know they are available on the host, I think we can either make the exclusive or not.

Thinking a bit on problems - when you set it the CPU `fixed` and `shared` (let's call it VM1). Once a VM with dedicated CPU starts (VM2), it might cause VM1 to lose the used CPUs as we re-assign them, right? I think we should document it as it might happen in such cases. If we are going to support the above logic.
This is the only problem I can think of by allowing it.

Reply:
According to 4. above, VMs with pinned CPUs are not touched on re-assignment (but non-pinned/dedicated VMs can be additionally placed on their CPUs). 
I don't know what all use cases there are for pinned CPUs but I can imagine one may want to have exclusive pinned CPUs.

Reply:
I was wrong in the previous comment. The conflict shouldn't happen. if the fixed CPU pinning is running (and we know these CPUs were available), they will run on them. Once you run a dedicated CPU VM, if there are not enough freed CPUs, it will fail. Otherwise it will use what available in the host. So conflict shouldn't happen.

Reply:
I think we can rule out dedicated with shared and think of it as the case of no-method + exclusive, no?

Reply:
> I think we can rule out dedicated with shared and think of it as the case of no-method + exclusive, no?

what I meant here is that when no assignment is defined (no method) and yet exclusiveness is requested then the CPUs will practically be "dedicated" (as they are pinned exclusively) and assigned by the logic that was described for the `dedicated` policy

other than that, lgtm

Reply:
After ruling out the uncertain combinations the only valid combination seems to be for CPU pinning with exclusive CPUs (no sharing). In my opinion tracking the exclusivity separately is unnecessary complication. The same can be achieved (from VDSM perspective) by adding a new value (e.g. `pin-exclusive`) to `cpuPolicy` enum in metadata.

This is an easy change and can be done any time later once we decide to implement this feature. I am going to resolve this conversation now.